### PR TITLE
Refactor/chart signals coords

### PR DESCRIPTION
**Pull request by GarageInc**
**Summary**

Move signals code into wrapper-HOC

Move some calculations into withXYCoords HOC

### CODE REVIEW

File: src/ducks/SANCharts/components/signalsChart/SignalsWrapper.js [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+            )
+        }
+      }
+    }
+
+    getLastPrice = () => {
+      const { chartData } = this.props
+
+      if (chartData.length === 0) {
+        return undefined
```
Comment (haritonasty) [inline src/ducks/SANCharts/components/signalsChart/SignalsWrapper.js]: skip undefined

File: src/ducks/SANCharts/components/rechartsWrapper/RechartsWrapper.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import React from 'react'
+import { Line } from 'recharts'
+import { chartBars } from '../../utils'
+import throttle from 'lodash.throttle'
```
Comment (haritonasty) [inline src/ducks/SANCharts/components/rechartsWrapper/RechartsWrapper.js]: rearrange imports

File: src/ducks/SANCharts/Charts.js [hunk 3]
```diff
@@ -43,18 +40,13 @@ import { checkHasPremium } from '../../pages/UserSelectors'
...
-  removeTrigger
-} from '../Signals/common/actions'
-import { buildPriceSignal } from '../Signals/utils/utils'
 import SignalLine, {
   SignalPointSvg
 } from './components/newSignalLine/SignalLine'
 import SidecarExplanationTooltip from './SidecarExplanationTooltip'
 import sharedStyles from './ChartPage.module.scss'
 import styles from './Chart.module.scss'
+import WithSignals from './components/signalsChart/SignalsWrapper'
```
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/Charts.js]: `styles` should be at the bottom of the imports

File: src/ducks/SANCharts/Charts.js [hunk 4]
```diff
@@ -112,110 +104,34 @@ class Charts extends React.Component {
...
+        'mousemove',
+        evt => onChartHover && onChartHover(evt, this.metricRef)
+      )
   }
 
   componentWillUnmount () {
     const chartSvg = this.props.chartRef.current
-    chartSvg && chartSvg.removeEventListener('mousemove', this.onChartHover)
+    const { onChartHover } = this.props
+    chartSvg &&
```
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/Charts.js]: `[D-1-1]` Conditional statement should be inside if block

File: src/ducks/SANCharts/Charts.js [hunk 5]
```diff
@@ -112,110 +104,34 @@ class Charts extends React.Component {
...
     const chartSvg = this.props.chartRef.current
-    chartSvg && chartSvg.addEventListener('mousemove', this.onChartHover)
-    this.loadSignals()
+    const { onChartHover } = this.props
+    chartSvg &&
+      chartSvg.addEventListener(
+        'mousemove',
+        evt => onChartHover && onChartHover(evt, this.metricRef)
+      )
   }
```
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/Charts.js]: Conditional statement should be inside `if` block

File: src/ducks/SANCharts/Charts.js [hunk 6]
```diff
@@ -112,110 +104,34 @@ class Charts extends React.Component {
...
-
   componentDidMount () {
     const chartSvg = this.props.chartRef.current
-    chartSvg && chartSvg.addEventListener('mousemove', this.onChartHover)
-    this.loadSignals()
+    const { onChartHover } = this.props
+    chartSvg &&
+      chartSvg.addEventListener(
+        'mousemove',
+        evt => onChartHover && onChartHover(evt, this.metricRef)
```
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/Charts.js]: `onChartHover` could be moved outside function inside `if` block

File: src/ducks/SANCharts/Charts.js [hunk 7]
```diff
@@ -112,110 +104,34 @@ class Charts extends React.Component {
...
   }
 
   componentWillUnmount () {
     const chartSvg = this.props.chartRef.current
-    chartSvg && chartSvg.removeEventListener('mousemove', this.onChartHover)
+    const { onChartHover } = this.props
+    chartSvg &&
+      chartSvg.removeEventListener(
+        'mousemove',
+        evt => onChartHover && onChartHover(evt, this.metricRef)
```
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/Charts.js]: `onChartHover` could be moved outside function inside `if` block

File: src/ducks/SANCharts/Charts.js [hunk 8]
```diff
@@ -374,6 +286,9 @@ class Charts extends React.Component {
     // HACK(vanguard): Because 'recharts' lib does not expose the "good" way to get coordinates
     this.xToYCoordinates = current.props[key] || []
 
+    const { setxToYCoordinates } = this.props
+    setxToYCoordinates && setxToYCoordinates(this.xToYCoordinates)
```
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/Charts.js]: Same as `[D-1-1]`

File: src/ducks/SANCharts/Charts.js [hunk 9]
```diff
@@ -636,7 +489,8 @@ class Charts extends React.Component {
             onMouseLeave={this.onMouseLeave}
             onMouseEnter={this.getXToYCoordinates}
             onMouseDown={event => {
-              this.onChartClick(event)
+              const { onChartClick } = this.props
+              onChartClick && onChartClick(event)
```
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/Charts.js]: Same as `[D-1-1]`

File: src/ducks/SANCharts/components/signalsChart/SignalsWrapper.js [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+} from './../../utils'
+import { Metrics } from './../../data'
+import {
+  createTrigger,
+  fetchSignals,
+  removeTrigger
+} from './../../../Signals/common/actions'
+import { buildPriceSignal } from './../../../Signals/utils/utils'
+
+const WithSignals = WrappedComponent => {
```
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/components/signalsChart/SignalsWrapper.js]: Higher order components should start with the lower case

File: src/ducks/SANCharts/components/signalsChart/SignalsWrapper.js [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+      const { isBeta, isLoggedIn, fetchSignals } = this.props
+
+      if (prevProps.isBeta !== isBeta || prevProps.isLoggedIn !== isLoggedIn) {
+        this.canShowSignalLines() && fetchSignals()
+      }
+    }
+
+    componentDidMount () {
+      const { fetchSignals } = this.props
+      this.canShowSignalLines() && fetchSignals()
```
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/components/signalsChart/SignalsWrapper.js]: Same as the `[D-1-1]`

File: src/ducks/SANCharts/components/signalsChart/SignalsWrapper.js [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+    render () {
+      const { signals = [], slug } = this.props
+
+      const { signalData } = this.state
+
+      const isSignalsEnabled = this.canShowSignalLines()
+
+      const { onSignalHover, onSignalLeave, onSignalClick } = this
+
+      const { xToYCoordinates } = this.state
```
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/components/signalsChart/SignalsWrapper.js]: State destruction also happens on the `LOC 171`

File: src/ducks/SANCharts/Charts.js (hunk 3 above)
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/Charts.js]: Why there is a `components` folder inside `SANCharts` duck-module? Everything there is already **component**. Tightly related components should be putted inside common component folder, e.g. `Signal`. Then, there will be consistency for folder names: `ducks/SANCharts/_.js`, `/ducks/SANCharts/Signal/withSignals.js`, `/ducks/SANCharts/Signal/SignalLine.js`

File: src/ducks/SANCharts/components/signalsChart/SignalsWrapper.js [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+      signalData: undefined,
+      signals: [],
+      xToYCoordinates: []
+    }
+
+    componentDidUpdate (prevProps, prevState, snapshot) {
+      const { isBeta, isLoggedIn, fetchSignals } = this.props
+
+      if (prevProps.isBeta !== isBeta || prevProps.isLoggedIn !== isLoggedIn) {
+        this.canShowSignalLines() && fetchSignals()
```
Comment (DmitriVanGuard) [inline src/ducks/SANCharts/components/signalsChart/SignalsWrapper.js]: ` this.canShowSignalLines()` should be listed inside the `if` block.
```js
if ( (prevProps.isBeta !== isBeta || prevProps.isLoggedIn !== isLoggedIn) && this.canShowSignalLines() ){
   /* .... */
}
```

**Pull request merged**